core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/fuse/PlanNodeFuser.java

```diff
@@ -0,0 +1,542 @@
+     *   i_size AS size
+     * FROM item
+     * M: brand2 → brand
+     * </pre>
+     * </p>
+     */
+    private Optional<FusedPlanNode> fuseTableScan(TableScanNode left, TableScanNode right)
+    {
+        if (!left.getTable().equals(right.getTable())) {
+            return Optional.empty();
+        }
+
+        // the fusing rules have to be run before predicate push-down into table scan
+        checkState(left.getEnforcedConstraint().isAll(), "Left node has non trivial enforcedConstraint %s, %s", left.getEnforcedConstraint(), left);
+        checkState(right.getEnforcedConstraint().isAll(), "Right node has non trivial enforcedConstraint %s, %s", right.getEnforcedConstraint(), right);
+        checkState(left.getUseConnectorNodePartitioning().equals(right.getUseConnectorNodePartitioning()),
+                "Fused table scans must have equal useConnectorNodePartitioning but got left %s in %s, right %s in %s",
+                left.getUseConnectorNodePartitioning(),
+                left, right.getUseConnectorNodePartitioning(),
+                right);
+        checkState(left.getStatistics().isEmpty(), "Left node has non empty statistics %s, %s", left.getStatistics(), left);
+        checkState(right.getStatistics().isEmpty(), "Right node has non trivial statistics %s, %s", right.getStatistics(), right);
+
+        if (!left.isUpdateTarget() == right.isUpdateTarget()) {
+            return Optional.empty();
+        }
+
+        Set<Symbol> outputs = new HashSet<>(left.getOutputSymbols());
+        Map<Symbol, ColumnHandle> assignments = new HashMap<>(left.getAssignments());
+        FusedSymbolMapping.Builder symbolMappingBuilder = FusedSymbolMapping.builder();
+        Set<Symbol> rightOutputSymbols = ImmutableSet.copyOf(right.getOutputSymbols());
+
+        // add unique assignments and outputs from the right side or map to the left side symbols
+        Map<ColumnHandle, Symbol> leftInverseAssignments = ImmutableBiMap.copyOf(left.getAssignments()).inverse();
+        right.getAssignments().forEach((rightSymbol, columnHandle) -> {
+            Symbol leftSymbol = leftInverseAssignments.get(columnHandle);
+            if (leftSymbol != null) {
+                // left assignment found, add the mapping and output symbol if necessary
+                symbolMappingBuilder.add(rightSymbol, leftSymbol);
+                if (rightOutputSymbols.contains(rightSymbol)) {
+                    outputs.add(leftSymbol);
+                }
+            }
+            else {
+                // left assignment not found, add a new assignment and output symbol if necessary
+                assignments.put(rightSymbol, columnHandle);
+                if (rightOutputSymbols.contains(rightSymbol)) {
+                    outputs.add(rightSymbol);
+                }
+            }
+        });
+        return Optional.of(new FusedPlanNode(
+                new TableScanNode(
+                        context.getIdAllocator().getNextId(),
```

Comment:
`left.getId`
We should be able to re-use `id` because if fuse is successful then the original `TableScanNode` would be discarded